src/Polysemy/Internal/Union.hs

```diff
@@ -256,51 +261,52 @@ tryMembership = tryMembership' @r @e
 -- | Decompose a 'Union'. Either this union contains an effect @e@---the head
 -- of the @r@ list---or it doesn't.
 decomp :: Union (e ': r) m a -> Either (Union r m a) (Weaving e m a)
-decomp (Union p a) =
+decomp (Union (UnionDetails p a)) =
   case p of
     Here  -> Right a
-    There pr -> Left $ Union pr a
+    There pr -> Left $ Union $ UnionDetails pr a
 {-# INLINE decomp #-}
 
 ------------------------------------------------------------------------------
 -- | Retrieve the last effect in a 'Union'.
 extract :: Union '[e] m a -> Weaving e m a
-extract (Union Here a)   = a
-extract (Union (There g) _) = case g of {}
+extract (Union (UnionDetails Here a))   = a
+extract (Union (UnionDetails (There g) _)) = case g of {}
 {-# INLINE extract #-}
 
 
 ------------------------------------------------------------------------------
 -- | An empty union contains nothing, so this function is uncallable.
 absurdU :: Union '[] m a -> b
-absurdU (Union pr _) = case pr of {}
+absurdU (Union (UnionDetails pr _)) = case pr of {}
 
 
 ------------------------------------------------------------------------------
 -- | Weaken a 'Union' so it is capable of storing a new sort of effect.
 weaken :: forall e r m a. Union r m a -> Union (e ': r) m a
-weaken (Union pr a) = Union (There pr) a
+weaken (Union (UnionDetails pr a)) = Union $ UnionDetails (There pr) a
 {-# INLINE weaken #-}
 
 
 
 ------------------------------------------------------------------------------
 -- | Lift an effect @e@ into a 'Union' capable of holding it.
-inj :: forall e r m a. (Functor m , Member e r) => e m a -> Union r m a
+inj :: forall e r rWoven a. (Member e r) => e (Sem rWoven) a -> Union r (Sem rWoven) a
```

Comment:
I like this variable naming better. I think I still prefer `r0` instead of `rWoven`, though, as `r0` better communicates that it represents the initial effect row, but I can be swayed.

Reply:
Perhaps rInitial to distinguish it in compiler errors from any variable names automatically assigned by the compiler by appending a number?

Reply:
Good idea.